File: java/src/main/java/com/tyouqu/database/NamedParameterSql.java
package com.tyouqu.database;

import java.util.ArrayList;
import java.util.List;

final class NamedParameterSql {
    private NamedParameterSql() {
    }

    static ParsedSql parse(String sql) {
        StringBuilder jdbcSql = new StringBuilder(sql.length());
        List<String> names = new ArrayList<>();
        boolean inSingleQuote = false;
        boolean inDoubleQuote = false;

        for (int i = 0; i < sql.length(); i++) {
            char current = sql.charAt(i);
            if (current == '\'' && !inDoubleQuote) {
                inSingleQuote = !inSingleQuote;
                jdbcSql.append(current);
                continue;
            }
            if (current == '"' && !inSingleQuote) {
                inDoubleQuote = !inDoubleQuote;
                jdbcSql.append(current);
                continue;
            }
            if (current == ':' && i + 1 < sql.length() && sql.charAt(i + 1) == ':') {
                jdbcSql.append("::");
                i++;
                continue;
            }
            if (current == ':' && !inSingleQuote && !inDoubleQuote && i + 1 < sql.length() && isNameStart(sql.charAt(i + 1))) {
                int start = i + 1;
                int end = start + 1;
                while (end < sql.length() && isNamePart(sql.charAt(end))) {
                    end++;
                }
                names.add(sql.substring(start, end));
                jdbcSql.append('?');
                i = end - 1;
                continue;
            }
            jdbcSql.append(current);
        }
        return new ParsedSql(jdbcSql.toString(), List.copyOf(names));
    }

    private static boolean isNameStart(char value) {
        return Character.isLetter(value) || value == '_';
    }

    private static boolean isNamePart(char value) {
        return Character.isLetterOrDigit(value) || value == '_';
    }

    record ParsedSql(String jdbcSql, List<String> parameterNames) {
    }
}

File: python/src/tyouqu_database/named_parameter_sql.py
from dataclasses import dataclass


@dataclass(frozen=True)
class ParsedSql:
    sql: str
    parameter_names: list[str]


def parse_named_parameter_sql(sql: str) -> ParsedSql:
    parsed = []
    names = []
    in_single_quote = False
    in_double_quote = False
    i = 0

    while i < len(sql):
        current = sql[i]
        if current == "'" and not in_double_quote:
            in_single_quote = not in_single_quote
            parsed.append(current)
            i += 1
            continue
        if current == '"' and not in_single_quote:
            in_double_quote = not in_double_quote
            parsed.append(current)
            i += 1
            continue
        if current == ":" and i + 1 < len(sql) and sql[i + 1] == ":":
            parsed.append("::")
            i += 2
            continue
        if (
            current == ":"
            and not in_single_quote
            and not in_double_quote
            and i + 1 < len(sql)
            and _is_name_start(sql[i + 1])
        ):
            start = i + 1
            end = start + 1
            while end < len(sql) and _is_name_part(sql[end]):
                end += 1
            names.append(sql[start:end])
            parsed.append("?")
            i = end
            continue
        parsed.append(current)
        i += 1

    return ParsedSql("".join(parsed), names)


def _is_name_start(value: str) -> bool:
    return value.isalpha() or value == "_"


def _is_name_part(value: str) -> bool:
    return value.isalnum() or value == "_"
